pkg/controller/migration/convert/network.go

```diff
@@ -98,22 +108,174 @@ func handleCalicoCNI(c *components, install *Installation) error {
 	}
+	if c.calicoCNIConfig.ContainerSettings.AllowIPForwarding {
+		return ErrIncompatibleCluster{"AllowIPForwarding not supported"}
+	}
 
 	return nil
 }
 
+func getNetworkingBackend(c *components) (string, error) {
+	netBackend, err := c.node.getEnv(ctx, c.client, containerCalicoNode, "CALICO_NETWORKING_BACKEND")
+	if err != nil {
+		return "", err
+	}
+
+	switch {
+	case netBackend == nil:
+		return "bird", nil
+	case *netBackend == "":
+		return "bird", nil
+	case *netBackend == "bird":
+		return "bird", nil
+	case *netBackend == "vxlan":
+		return "vxlan", nil
+	case *netBackend == "none":
+		return "none", nil
+	default:
+		return "", fmt.Errorf("CALICO_NETWORKING_BACKEND %s is not valid", *netBackend)
+	}
+}
+
+func subhandleCalicoIPAM(c *components, install *Installation) error {
+	if c.calicoCNIConfig.IPAM.Type != "calico-ipam" {
+		// I don't know if this should be an error or not, the way it is written
+		// this check is just double checking what the caller should have already
+		// validated.
+		return nil
+	}
+
+	// TODO: Is there anything we need to do with the Name field on the CNI plugin or IPAM
+	// plugin? We obviously don't have a way currently to persist them.
+
+	netBackend, err := getNetworkingBackend(c)
+	if err != nil {
+		return err
+	}
+
+	switch netBackend {
+	case "bird":
+		install.Spec.CalicoNetwork.BGP = operatorv1.BGPOptionPtr(operatorv1.BGPEnabled)
+	case "vxlan":
+		install.Spec.CalicoNetwork.BGP = operatorv1.BGPOptionPtr(operatorv1.BGPDisabled)
+	default:
+		return ErrIncompatibleCluster{fmt.Sprintf("CALICO_NETWORKING_BACKEND %s is not valid", netBackend)}
+	}
+
+	invalidFields := []string{}
+	if c.calicoCNIConfig.IPAM.Subnet != "" {
+		invalidFields = append(invalidFields, "ipam.subnet field is unsupported")
+	}
+
+	if len(c.calicoCNIConfig.IPAM.IPv4Pools) != 0 {
+		invalidFields = append(invalidFields, "ipam.ipv4pools field is unsupported")
+	}
+	if len(c.calicoCNIConfig.IPAM.IPv6Pools) != 0 {
+		invalidFields = append(invalidFields, "ipam.ipv6pools field is unsupported")
+	}
+
+	if len(invalidFields) > 0 {
+		return ErrIncompatibleCluster{"configuration could not be migrated: " + strings.Join(invalidFields, ",")}
+	}
+	return nil
+}
+
+const UnsupportedMsg = "is unsupported"
+
+// subhandleHostLocalIPAM checks all fields in the Host Local IPAM configuration,
+// if any fields have unexpected values an error message will be returned.
+// The function tries to collect all the errors and report one message.
+// If there are no errors and the config can be added to the 'install' passed
+// in then nil is returned.
+func subhandleHostLocalIPAM(c *components, install *Installation) error {
+	if c.calicoCNIConfig.IPAM.Type != "host-local" {
+		// I don't know if this should be an error or not, the way it is written
+		// this check is just double checking what the caller should have already
+		// validated.
+		return nil
+	}
+
+	if c.hostLocalIPAMConfig == nil {
+		return ErrIncompatibleCluster{"no configuration"}
```

Comment:
What's the scenario where this error would be hit?  I think it's when they're using host-local, but for some reason, `getComponents` didn't initialize `c.hostLocalIPAMConfig` with the parsed config. The only situation where that woudl occur is probably developer error? Maybe CNI parsing for host-local should just be done here in the host-local handler instead of during `getComponents` :shrug:

Reply:
I think it would be a developer error. Because when processing the CNI if IPAM.Type is hostlocal then a hostLocalConfig should be parsed and if that failed it should be an error there, which should block us from ever getting to this point.

I did consider doing that parsing here but thought it was better to keep those 2 jobs separated. With the loadCNI loading the config into our structs and then just processing those structs here.